race-buffer/src/lib.rs

```diff
@@ -1,162 +1,234 @@
+    if use_base_2_indexing {
+        // Index is lowest n bits of seqn where storage_cap is 2^n
+        (Into::<u64>::into(seqn) & (storage_cap as u64 - 1)) as usize
+    } else {
+        (Into::<u64>::into(seqn) % storage_cap as u64) as usize
+    }
+}
 
-        // Invariant: Check if entries all have correct index
-        fn entries_correct_index(rbuf: &[Option<OrderedEntry>]) -> bool {
-            for (idx, entry) in rbuf.iter().enumerate().filter(|(_, e)| e.is_some()) {
-                if entry.unwrap().to_index() != idx as u32 {
-                    return false;
-                }
-            }
-            return true;
-        }
+/// Calculate the number of entries missed based on the read and overwrite sequence numbers
+#[inline]
+fn num_missed(read_seqn: SeqNum, overwrite_seqn: SeqNum) -> SeqNum {
+    if overwrite_seqn < read_seqn {
+        0.into()
+    } else {
+        overwrite_seqn - read_seqn
+    }
+}
 
-        // Invariant: Check if entries all have correct index
-        fn double_entries_consistent(rbuf: &[Option<OrderedEntry>]) -> bool {
-            if rbuf.len() == 0 {
-                return true;
-            }
-            if let Some(first_entry) = rbuf[0] {
-                if first_entry.is_suffix() {
-                    return false;
-                }
-            }
-            if let Some(last_entry) = rbuf.last().unwrap() {
-                if last_entry.is_prefix_unchecked() {
-                    return false;
-                }
-            }
-            for i in 0..(rbuf.len() - 1) {
-                match rbuf[i] {
-                    None => {
-                        if let Some(next) = rbuf[i + 1] {
-                            if next.is_suffix() {
-                                return false;
-                            }
-                        }
-                    }
-                    Some(current) => {
-                        if current.is_prefix_unchecked() {
-                            if let Some(next) = rbuf[i + 1] {
-                                if !next.is_suffix() {
-                                    return false;
-                                }
-                            } else {
-                                return false;
-                            }
-                        }
-                    }
-                }
-            }
-            return true;
+/// Entry that can be stored in a RaceBuffer
+pub trait Entry: Copy + PartialEq {
+    /// Return true if entry is the first in a double entry
+    fn is_prefix(&self) -> bool;
+}
+
+/// An entry or double entry that has just been overwritten
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum WholeEntry<E>
+where
+    E: Entry,
+{
+    /// Single entry overwritten
+    Single(E),
+    /// Double entry overwritten
+    Double(E, E),
+}
+
+impl<E> WholeEntry<E>
+where
+    E: Entry,
+{
+    /// 1 if entry is single, 2 if double
+    pub fn size(&self) -> u64 {
```

Comment:
If the value is only 1 or 2, I'd recommend using a smaller type for the output, perhaps a u8.